### PR TITLE
Implement the `BucketQueueView`

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

The `QueueView` is used for the protocol and this leads to painful-to-read loops where we read the front, then deleting and so on until we get what we want. This leads to many storage calls which are not storage friendly.

## Proposal

The following change was proposed:
* It was identified that only the `front`, `delete_front`, and `push_back` are used in the protocol. So, this is what needs to be optimized.
* The arrangement of keys is made so that when the view is loaded the front is loaded as well. 
* So, the `BucketQueueView` has `fn front` being a regular function. But of course, it has a cost in that the `delete_front` needs to be async and can return an error.
* All the other functions of `QueueView` are implemented in `BucketQueueView`.

The protocol has not been changed so far but could be later on. Whether we keep `QueueView` in the future is an open question.

## Test Plan

The unit test have been added to the documentation of the code. The random tests are added. Also benchmarks are added.
The benchmarks proved the reliability of the idea. Two realistic scenarios are created and run against `QueueView` and `BucketQueueView`. The result is that `QueueView<_,u8>` takes 836 microseconds and `BucketQueueView<_,u8,100>` takes 177 microseconds.

## Release Plan

Not relevant.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

**Comment by ma2bd** [comment]
What is the result of the micro benchmarks (and other available measurements) when comparing `BucketQueueView` and `QueueView`?

**Comment by MathieuDutSik** [comment]
> What is the result of the micro benchmarks (and other available measurements) when comparing `BucketQueueView` and `QueueView`?

As indicated above, 836 microseconds for the QueueView and 177 microseconds for the BucketQueueView.

File: linera-views/src/views/bucket_queue_view.rs [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+        "BucketQueueView hash runtime",
+        &[],
+        Some(vec![
+            0.001, 0.003, 0.01, 0.03, 0.1, 0.2, 0.3, 0.4, 0.5, 0.75, 1.0, 2.0, 5.0,
+        ]),
+    )
+    .expect("Histogram can be created")
+});
+
+/// Key tags to create the sub-keys of a BucketQueueView on top of the base key.
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:41]:
```suggestion
/// Key tags to create the sub-keys of a [`BucketQueueView`] on top of the base key.
```

File: linera-views/src/views/bucket_queue_view.rs [hunk 2]
```diff
@@ -0,0 +1,656 @@
...
+#[repr(u8)]
+enum KeyTag {
+    /// Prefix for the front of the view
+    Front = MIN_VIEW_TAG,
+    /// Prefix for the storing of the stored-indices information
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:46]:
```suggestion
    /// Prefix for [`StoredIndices`]
```

or

```suggestion
    /// Prefix for the list of indices
```

File: linera-views/src/views/bucket_queue_view.rs [hunk 3]
```diff
@@ -0,0 +1,656 @@
...
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+struct StoredIndices {
+    indices: Vec<(usize, usize)>,
+    position: usize,
+}
+
+impl StoredIndices {
+    fn is_empty(&self) -> bool {
+        self.indices.len() == 0
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:60]:
```suggestion
        self.indices.is_empty()
```

(I'm surprised CI passed: https://rust-lang.github.io/rust-clippy/v0.0.212/#len_zero)

File: linera-views/src/views/bucket_queue_view.rs [hunk 4]
```diff
@@ -0,0 +1,656 @@
...
+        })
+    }
+}
+
+impl<C, T, const N: usize> BucketQueueView<C, T, N>
+where
+    C: Context + Send + Sync,
+    ViewError: From<C::Error>,
+{
+    /// Get the key corresponding to the index
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:266]:
```suggestion
    /// Gets the key corresponding to the index
```

(Also below.)

File: linera-views/src/views/bucket_queue_view.rs [hunk 5]
```diff
@@ -0,0 +1,656 @@
...
+    fn get_index_key(&self, index: usize) -> Result<Vec<u8>, ViewError> {
+        Ok(if index == 0 {
+            self.context.base_tag(KeyTag::Front as u8)
+        } else {
+            self.context.derive_tag_key(KeyTag::Index as u8, &index)?
+        })
+    }
+
+    /// Get the stored_count
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:275]: That comment isn't helpful.

File: linera-views/src/views/bucket_queue_view.rs [hunk 6]
```diff
@@ -0,0 +1,656 @@
...
+    Store,
+    /// Prefix for the indices of the log.
+    Index,
+}
+
+#[derive(Clone, Debug, Default, Serialize, Deserialize)]
+struct StoredIndices {
+    indices: Vec<(usize, usize)>,
+    position: usize,
+}
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:56]: Please document this: what indices are these, and the position of what in what?

Looking at the code: Is the first `usize` the number of entries in a bucket, and the second one the index of the first item in that bucket?

Wouldn't it be easier to make all buckets the same size, i.e. if we add elements we add them to the last bucket until it's full? Then we wouldn't need a list of indices at all?

File: linera-views/src/views/bucket_queue_view.rs [hunk 7]
```diff
@@ -0,0 +1,656 @@
...
+        }
+    }
+
+    pub fn is_incrementable(&self) -> bool {
+        self.position.is_some()
+    }
+}
+
+/// A view that supports a FIFO queue for values of type `T`.
+pub struct BucketQueueView<C, T, const N: usize> {
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:86]: Is `N` the bucket size? Does this really need to be a const generic rather than a field? Won't this generate new code for each value of `N` we use?

**Review by afck — COMMENTED** [review]
Sorry, I'm having trouble reviewing this: Could you document the semantics and design of this view?

And should we remove the `QueueView`, or are there use cases where that's preferable over a `BucketQueueView`?

File: linera-views/src/views/bucket_queue_view.rs (hunk 6 above)
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:56]: The list of indices cannot be all of the same size. Suppose it that way and work out the argument:
* Suppose for the sake of argument that N=5.
* Suppose that the list has 13 elements. But supposed that we have just read the front (of size 5).
* Suppose that we have 2 additional elements in `new_back_values` and that we want to write them in the flush.
* Then the flush operation needs to create a list of 5 elements with 3 from the storage and 2 from the `new_back_values`. 
* So, the flush needs to access the database, which is not possible since flush is not async.

So, yes, the code tries to have the buckets all the same size, but that is not always possible.

File: linera-views/src/views/bucket_queue_view.rs (hunk 7 above)
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:86]: Yes, `N` is the bucket size. And yes, code will be generated for each instance used in the code. I do not know the details of the code generation in Rust, but I know better how things go in C++. It ought to be similar.

The recommendation in C++ is that there is no downside to having many specific type instances. In any case, the planned usage are about 5-10 instances, so I do not think that it will be a problem.

I do not see how we could have a bucket size that is a field. That size should turn out from the `Self::load` function. If the view is empty, then we cannot get the size.

File: linera-views/src/context.rs [hunk 8]
```diff
@@ -167,11 +167,11 @@ pub trait Context: Clone {
 #[derive(Debug, Default, Clone)]
 pub struct ViewContext<E, S> {
     /// The DB client that is shared between views.
-    store: S,
+    pub store: S,
```
Comment (ma2bd) [inline linera-views/src/context.rs:170]: Why is this needed? I just made a PR to avoid this

**Comment by ma2bd** [comment]
> > What is the result of the micro benchmarks (and other available measurements) when comparing `BucketQueueView` and `QueueView`?
> 
> As indicated above, 836 microseconds for the QueueView and 177 microseconds for the BucketQueueView.

Thanks, this is promising indeed. Can you make it clear in the summary which command takes 856ms vs 177ms? Someone must be able to reproduce the experiment without taking a guess.

**Comment by MathieuDutSik** [comment]
> > > What is the result of the micro benchmarks (and other available measurements) when comparing `BucketQueueView` and `QueueView`?
> > 
> > 
> > As indicated above, 836 microseconds for the QueueView and 177 microseconds for the BucketQueueView.
> 
> Thanks, this is promising indeed. Can you make it clear in the summary which command takes 856ms vs 177ms? Someone must be able to reproduce the experiment without taking a guess.

The number are available by running
```
cargo bench memory_queue_view
```

and

```
cargo bench memory_bucket_queue_view
```

File: linera-views/src/context.rs (hunk 8 above)
Comment (MathieuDutSik) [reply linera-views/src/context.rs:170]: Ok, I reverted the change and changed the way the `ViewContext` is build.

File: linera-views/src/views/bucket_queue_view.rs [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+            self.stored_indices.indices.drain(0..i_block);
+            self.stored_data.drain(0..i_block);
+            self.cursor = Cursor {
+                position: Some((0, position)),
+            };
+            // We need to ensure that the first index is in the front.
+            let first_index = self.stored_indices.indices[0].1;
+            if first_index != 0 {
+                let size = self.stored_indices.indices[0].0;
+                self.stored_indices.indices[0] = (size, 0);
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:218]: But why do we bother at all enforcing that the first index is always 0?

Why not instead keep the indices consecutive? Then we also wouldn't have to store each index, only the first one?

File: linera-views/src/views/bucket_queue_view.rs [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+            }
+        }
+        if !self.new_back_values.is_empty() {
+            delete_view = false;
+            let mut unused_index = match self.stored_indices.indices.last() {
+                Some((_, index)) => index + 1,
+                None => 0,
+            };
+            let new_back_values: VecDeque<T> = std::mem::take(&mut self.new_back_values);
+            let new_back_values: Vec<T> = new_back_values.into_iter().collect::<Vec<_>>();
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:233]: There's also `VecDeque::make_consecutive`, which doesn't reallocate.

File: linera-views/src/views/bucket_queue_view.rs [hunk 11]
```diff
@@ -0,0 +1,700 @@
...
+    /// Gets the key corresponding to the index
+    fn get_index_key(&self, index: usize) -> Result<Vec<u8>, ViewError> {
+        Ok(if index == 0 {
+            self.context.base_tag(KeyTag::Front as u8)
+        } else {
+            self.context.derive_tag_key(KeyTag::Index as u8, &index)?
+        })
+    }
+
+    /// Get the number of entries in the container that are stored
```
Comment (afck) [inline linera-views/src/views/bucket_queue_view.rs:297]:
```suggestion
    /// Gets the number of entries in the container that are stored
```

**Review by afck — COMMENTED** [review]
We often clone values instead of returning a reference, and there are some `&self` (non-mutable) functions that load from storage and then don't update the in-memory view.

Would it be better to return references where feasible?

And should we use interior mutability to update the loaded view whenever we load something from storage, and make a note whether the loaded value is still known to match the stored one?

File: linera-views/src/views/bucket_queue_view.rs [hunk 12]
```diff
@@ -0,0 +1,700 @@
...
+    async fn load_all(&mut self) -> Result<(), ViewError> {
+        if !self.delete_storage_first {
+            let elements = self.elements().await?;
+            self.new_back_values.clear();
+            for elt in elements {
+                self.new_back_values.push_back(elt);
+            }
+            self.cursor = Cursor { position: None };
+            self.delete_storage_first = true;
+        }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:614]: Yes.

File: linera-views/src/views/bucket_queue_view.rs (hunk 11 above)
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:297]: Ok.

File: linera-views/src/views/bucket_queue_view.rs [hunk 13]
```diff
@@ -0,0 +1,700 @@
...
+    /// queue.push_back(34);
+    /// assert_eq!(queue.count(), 1);
+    /// # })
+    /// ```
+    pub fn count(&self) -> usize {
+        self.stored_count() + self.new_back_values.len()
+    }
+}
+
+impl<'a, C, T, const N: usize> BucketQueueView<C, T, N>
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:342]: For the `IterMut<'a, T>`

File: linera-views/src/views/bucket_queue_view.rs [hunk 14]
```diff
@@ -0,0 +1,700 @@
...
+    /// # tokio_test::block_on(async {
+    /// # use linera_views::context::create_test_memory_context;
+    /// # use linera_views::bucket_queue_view::BucketQueueView;
+    /// # use crate::linera_views::views::View;
+    /// # let context = create_test_memory_context();
+    /// let mut queue = BucketQueueView::<_, u128, 5>::load(context).await.unwrap();
+    /// queue.push_back(34);
+    /// queue.push_back(37);
+    /// queue.push_back(47);
+    /// assert_eq!(queue.read_front(2).await.unwrap(), vec![34, 37]);
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:575]: Thanks.

File: linera-views/src/views/bucket_queue_view.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-views/src/views/bucket_queue_view.rs:342]: But then it could be on the function that uses it, instead of the whole `impl` block?

File: linera-views/src/views/bucket_queue_view.rs [hunk 15]
```diff
@@ -0,0 +1,700 @@
...
+    }
+
+    async fn hash(&self) -> Result<<Self::Hasher as Hasher>::Output, ViewError> {
+        #[cfg(with_metrics)]
+        let _hash_latency = BUCKET_QUEUE_VIEW_HASH_RUNTIME.measure_latency();
+        let elements = self.elements().await?;
+        let mut hasher = sha3::Sha3_256::default();
+        hasher.update_with_bcs_bytes(&elements)?;
+        Ok(hasher.finalize())
+    }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:659]: There are some possibilities for improving the hashing. However, in my opinion, this is not urgent and instead, the issues of storage have to be the top priority.

File: linera-views/src/views/bucket_queue_view.rs [hunk 16]
```diff
@@ -0,0 +1,700 @@
...
+            if i_block != 0 || position != self.stored_indices.position {
+                return true;
+            }
+        }
+        !self.new_back_values.is_empty()
+    }
+
+    fn flush(&mut self, batch: &mut Batch) -> Result<bool, ViewError> {
+        let mut delete_view = false;
+        assert_eq!(self.stored_data.len(), self.stored_indices.len());
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:192]: Yes, that makes sense. Thank you.

File: linera-views/src/views/bucket_queue_view.rs [hunk 17]
```diff
@@ -0,0 +1,700 @@
...
+    /// # use linera_views::bucket_queue_view::BucketQueueView;
+    /// # use crate::linera_views::views::View;
+    /// # let context = create_test_memory_context();
+    /// let mut queue = BucketQueueView::<_, u8, 5>::load(context).await.unwrap();
+    /// queue.push_back(34);
+    /// queue.push_back(42);
+    /// assert_eq!(queue.front(), Some(34));
+    /// # })
+    /// ```
+    pub fn front(&self) -> Option<T> {
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:361]: You are quite right and that is indeed another small benefit of this container: We can remove some clone operations.

File: linera-views/src/views/bucket_queue_view.rs [hunk 18]
```diff
@@ -0,0 +1,700 @@
...
+    /// let mut iter = queue.iter_mut().await.unwrap();
+    /// let value = iter.next().unwrap();
+    /// *value = 42;
+    /// assert_eq!(queue.elements().await.unwrap(), vec![42]);
+    /// # })
+    /// ```
+    pub async fn iter_mut(&'a mut self) -> Result<IterMut<'a, T>, ViewError> {
+        self.load_all().await?;
+        Ok(self.new_back_values.iter_mut())
+    }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:636]: Yes, that is not great I agree. The problem is that the iterator is not async, so data has to be downloaded at once and then worked on.

File: linera-views/src/views/bucket_queue_view.rs (hunk 9 above)
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:218]: Yes, we want to preload the data right away.
Yes, that code has been corrected away.

[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:218]: Yes, that has been corrected. It is now obsolete code.

File: linera-views/src/views/bucket_queue_view.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-views/src/views/bucket_queue_view.rs:636]: It could still be read all at once when creating the iterator.
But we could keep track of which values were actually _returned_ by the iterator: The ones that weren't can't have changed, so they wouldn't need to be rewritten when saving the view.

File: linera-views/src/views/bucket_queue_view.rs [hunk 19]
```diff
@@ -0,0 +1,700 @@
...
+                    position + size
+                };
+                for element in &vec[position..end] {
+                    elements.push(element.clone());
+                }
+                if size >= count_remain {
+                    count_remain = 0;
+                    break;
+                }
+                count_remain -= size;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:535]: I think that we need to keep track. That lets us know when to stop in the loop.

File: linera-views/src/views/bucket_queue_view.rs [hunk 20]
```diff
@@ -0,0 +1,700 @@
...
+                }
+                if size >= count_remain {
+                    break;
+                }
+                count_remain -= size;
+                position = 0;
+            }
+            let values = self.context.read_multi_values_bytes(keys).await?;
+            position = pair.1;
+            let mut pos = 0;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:511]: I prefer to use `value_pos` here.

File: linera-views/src/views/bucket_queue_view.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (afck) [reply linera-views/src/views/bucket_queue_view.rs:535]: But we need to stop whenever `count == elements.len()`? No `count_remain` needed?

File: linera-views/src/views/bucket_queue_view.rs [hunk 21]
```diff
@@ -0,0 +1,700 @@
...
+        let Some((len, index)) = self.stored_indices.indices.last() else {
+            return Ok(None);
+        };
+        if let Some(vec) = self.stored_data.last().unwrap() {
+            return Ok(Some(vec.last().unwrap().clone()));
+        }
+        let key = self.get_index_key(*index)?;
+        let value = self.context.read_value_bytes(&key).await?;
+        let value = value.as_ref().ok_or(ViewError::MissingEntries)?;
+        let value = bcs::from_bytes::<Vec<T>>(value)?;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs]: Ok, done.

File: linera-views/src/views/bucket_queue_view.rs [hunk 22]
```diff
@@ -0,0 +1,700 @@
...
+        self.new_back_values.clear();
+    }
+
+    async fn has_pending_changes(&self) -> bool {
+        if self.delete_storage_first {
+            return true;
+        }
+        if !self.stored_indices.is_empty() {
+            let Some((i_block, position)) = self.cursor.position else {
+                return true;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:212]: The line is reachable if there is some data in the `stored_data` (first test `is_empty`).
Then the `return true` is reached if the cursor position is `None`. That position can be reached if enough data is deleted.

File: linera-views/src/views/bucket_queue_view.rs [hunk 23]
```diff
@@ -0,0 +1,700 @@
...
+        &[],
+        Some(vec![
+            0.001, 0.003, 0.01, 0.03, 0.1, 0.2, 0.3, 0.4, 0.5, 0.75, 1.0, 2.0, 5.0,
+        ]),
+    )
+    .expect("Histogram can be created")
+});
+
+/// Key tags to create the sub-keys of a [`BucketQueueView`] on top of the base key.
+/// * The Front is special and directly accessible so that the
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:42]: Corrected.

File: linera-views/src/views/bucket_queue_view.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:370]: Ok, we can do it that way.

File: linera-views/src/views/bucket_queue_view.rs [hunk 24]
```diff
@@ -0,0 +1,700 @@
...
+    pub async fn read_back(&self, count: usize) -> Result<Vec<T>, ViewError> {
+        if count <= self.new_back_values.len() {
+            let start = self.new_back_values.len() - count;
+            Ok(self
+                .new_back_values
+                .range(start..)
+                .cloned()
+                .collect::<Vec<_>>())
+        } else {
+            let mut increment = self.count() - count;
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:648]: Ok, I corrected that.

File: linera-views/src/views/bucket_queue_view.rs [hunk 25]
```diff
@@ -0,0 +1,700 @@
...
+
+    #[async_graphql::Object(cache_control(no_cache), name_type)]
+    impl<C: Context, T: async_graphql::OutputType, const N: usize> BucketQueueView<C, T, N>
+    where
+        C: Send + Sync,
+        T: serde::ser::Serialize + serde::de::DeserializeOwned + Clone + Send + Sync,
+    {
+        async fn entries(&self, count: Option<usize>) -> async_graphql::Result<Vec<T>> {
+            Ok(self
+                .read_front(count.unwrap_or_else(|| self.count()))
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:757]: Yes, this has been corrected.
However, the proposed change would not work because `Option<usize>` is not the format of the input.

File: linera-views/src/views/bucket_queue_view.rs [hunk 26]
```diff
@@ -0,0 +1,700 @@
...
+            self.stored_indices = StoredIndices::default();
+            self.stored_data.clear();
+        } else if let Some((i_block, position)) = self.cursor.position {
+            for block in 0..i_block {
+                let index = self.stored_indices.indices[block].1;
+                let key = self.get_index_key(index)?;
+                batch.delete_key(key);
+            }
+            self.stored_indices.indices.drain(0..i_block);
+            self.stored_data.drain(0..i_block);
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs]: Ok, done.

File: linera-views/src/views/bucket_queue_view.rs (hunk 10 above)
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:264]: There is no function `make_consecutive`. There is a `make_contiguous` but I could not get it to work.
Note that this code does not clone.

File: linera-views/src/views/bucket_queue_view.rs (hunk 18 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:697]: I had this kind of issues for the `QueueView`. Yes, I think there is a possibility of improvement. However, I do not see it as critical.

File: linera-views/src/views/bucket_queue_view.rs (hunk 19 above)
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:598]: I am not sure what is the issue here. For sure we could rename the variable or change sign.
But in the end of the loop, we need to extract the relevant entries in the `new_back_values`. And the size needed is the `count_remain`.

File: linera-views/src/views/bucket_queue_view.rs [hunk 27]
```diff
@@ -0,0 +1,700 @@
...
+            let start = self.new_back_values.len() - count;
+            Ok(self
+                .new_back_values
+                .range(start..)
+                .cloned()
+                .collect::<Vec<_>>())
+        } else {
+            let mut increment = self.count() - count;
+            let Some((i_block, mut position)) = self.cursor.position else {
+                unreachable!();
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:650]: If the unreachable were called, then it would mean that the `cursor.position = None`. In that case, you would get that `self.count() == self.new_back_values.len()`. But that case would be covered by the first if loop.

File: linera-views/src/views/bucket_queue_view.rs [hunk 28]
```diff
@@ -0,0 +1,700 @@
...
+                    }
+                };
+                let end = if count_remain <= size {
+                    position + count_remain
+                } else {
+                    position + size
+                };
+                for element in &vec[position..end] {
+                    elements.push(element.clone());
+                }
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/views/bucket_queue_view.rs:594]: Thank you. That indeed simplified the code.

**Comment by MathieuDutSik** [comment]
> And should we use interior mutability to update the loaded view whenever we load something from storage, and make a note whether the loaded value is still known to match the stored one?

Thank you for your review. I considered interior mutability. I decided against as the use case where we want to use the `BucketQueueView` this would not change anything.

File: linera-views/src/views/bucket_queue_view.rs (hunk 10 above)
Comment (afck) [reply linera-views/src/views/bucket_queue_view.rs]: (Sorry, that's what I meant. No worries, maybe I'll give it a try later. Not important.)

File: linera-views/benches/queue_view.rs [hunk 29]
```diff
@@ -0,0 +1,227 @@
...
+{
+    let context = ViewContext::<(), S>::create_root_context(store, ())
+        .await
+        .unwrap();
+    let mut total_time = Duration::ZERO;
+    let mut rng = make_deterministic_rng();
+    for _ in 0..iterations {
+        let operations = generate_test_case(N_OPERATIONS, &mut rng);
+        let mut view = QueueStateView::load(context.clone()).await.unwrap();
+        //
```
Comment (afck) [inline linera-views/benches/queue_view.rs:76]:
```suggestion
```

File: linera-views/tests/random_container_tests.rs [hunk 30]
```diff
@@ -402,6 +403,113 @@ async fn map_view_mutability() -> Result<()> {
...
+    let mut rng = test_utils::make_deterministic_rng();
+    let mut vector = Vec::new();
+    let n = 200;
+    for _ in 0..n {
+        let mut view = BucketQueueStateView::load(context.clone()).await?;
+        let hash = view.crypto_hash().await?;
+        let save = rng.gen::<bool>();
+        let elements = view.queue.elements().await?;
+        assert_eq!(elements, vector);
+        //
```
Comment (afck) [inline linera-views/tests/random_container_tests.rs:423]:
```suggestion
```

**Review by afck — APPROVED** [review]
